parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+                return cands
+        # if we aren't training, return label candidates if available
+        return obs.get('label_candidates')
+
+    def _init_embeddings(self, log=True):
+        """Copy embeddings from the pretrained embeddings to the lookuptable.
+
+        :param weight:   weights of lookup table (nn.Embedding/nn.EmbeddingBag)
+        :param emb_type: pretrained embedding type
+        """
+        weight = self.model.lt.weight
+        emb_type = self.opt.get('embedding_type', 'random')
+        if emb_type == 'random':
+            return
+        embs, name = self._get_embtype(self, emb_type)
+        cnt = 0
+        for w, i in self.dict.tok2ind.items():
+            if w in embs.stoi:
+                vec = self._project_vec(self,
+                                        embs.vectors[embs.stoi[w]],
+                                        weight.size(1))
+                weight.data[i] = vec
+                cnt += 1
+        if log:
+            print('Initialized embeddings for {} tokens ({}%) from {}.'
+                  ''.format(cnt, round(cnt * 100 / len(self.dict), 1), name))
+
+    def same(self, labels, cand):
+        if labels:
+            for label in labels:
+                if label == cand:
+                    return True
+        return False
+
+    def input_dropout(self, xs, ys, negs):
+        """Dropout some input elements"""
+
+        def dropout(x, rate):
+            x_len = x.size(1)
+            num_keep = x_len - int(rate * x_len)
+            idxs = sorted(random.sample(list(range(x_len)), num_keep))
+            if self.use_cuda:
+                to_keep = torch.LongTensor(idxs).cuda()
+            else:
+                to_keep = torch.LongTensor(idxs)
+            new_x = x[:, to_keep]
+            return new_x
+
+        rate = self.opt.get('input_dropout')
+
+        # NOTE: we drop out uniformly along the batch so that we don't need
+        # to perform padding again
+        xs2 = dropout(xs, rate)
+        ys2 = dropout(ys, rate)
+        negs2 = dropout(negs.view(-1, negs.size(-1)), rate)
+        negs2 = negs2.view(negs.size(0), negs.size(1), -1)
+
+        return xs2, ys2, negs2
+
+    def add_to_ys_cache(self, ys):
+        if not ys:
+            return
+        if len(self.ys_cache) < self.ys_cache_sz:
+            self.ys_cache.append(ys)
+        else:
+            ind = random.randint(0, self.ys_cache_sz - 1)
+            self.ys_cache[ind] = ys
+
+    def optimizer_reset(self):
+        lr = self.opt['learningrate']
+        optim_class = StarspaceTorchAgent.OPTIM_OPTS[self.opt['optimizer']]
+        kwargs = {'lr': lr}
+        self.optimizer = optim_class(self.model.parameters(), **kwargs)
+
+    def reset(self):
+        """Reset observation and episode_done."""
+        self.observation = {}
+        self.history.clear()
+        self.replies.clear()
+        self.reset_metrics()
+        # reset optimizer
+        self.optimizer_reset()
+
+    def share(self):
```

Comment:
Both dict and model are set in TorchAgent and TorchRankerAgent respectively, so I don't think you need this.